src/crimson/os/seastore/extent_pinboard.cc

```diff
@@ -0,0 +1,847 @@
+
+    laddr_t laddr;
+    extent_len_t length;
+    extent_len_t last_access_end;
+  };
+
+  using entry_queue_t = std::list<entry_t>;
+  using entry_index_t = boost::unordered_flat_map<
+    laddr_t, entry_queue_t::iterator>;
+
+  void remove(entry_index_t::iterator iter) {
+    assert(iter != index.end());
+    assert(iter->second != queue.end());
+    assert(current_size >= iter->second->length);
+    current_size -= iter->second->length;
+    queue.erase(iter->second);
+    index.erase(iter);
+  }
+
+  void trim_to(size_t target) {
+    while (current_size > target) {
+      assert(!queue.empty());
+      assert(queue.size() == index.size());
+      remove(index.find(queue.front().laddr));
+    }
+    if (target == 0) {
+      assert(current_size == 0);
+      assert(queue.empty());
+      assert(index.empty());
+    }
+  }
+
+  const size_t capacity;
+  size_t current_size;
+  entry_queue_t queue;
+  entry_index_t index;
+};
+
+class ExtentPinboardTwoQ : public ExtentPinboard {
+public:
+  ExtentPinboardTwoQ(
+    size_t warm_in_capacity,
+    size_t warm_out_capacity,
+    size_t hot_capacity)
+      : warm_in(warm_in_capacity),
+	warm_out(warm_out_capacity),
+	hot(hot_capacity)
+  {
+    LOG_PREFIX(ExtentPinboardTwoQ::ExtentPinboardTwoQ);
+    INFO("created, warm_in_capacity=0x{:x}B, "
+	 "warm_out_capacity=0x{:x}B, hot_capacity=0x{:x}B",
+	 warm_in_capacity, warm_out_capacity, hot_capacity);
+  }
+
+  size_t get_capacity_bytes() const {
+    return warm_in.get_capacity_bytes() + hot.get_capacity_bytes();
+  }
+
+  size_t get_current_size_bytes() const final {
+    return warm_in.get_current_size_bytes() + hot.get_current_size_bytes();
+  }
+
+  size_t get_current_num_extents() const final {
+    return warm_in.get_current_num_extents() + hot.get_current_num_extents();
+  }
+
+  void register_metrics() final;
+
+  void get_stats(
+    cache_stats_t &stats,
+    bool report_detail,
+    double seconds) const final;
+
+  void remove(CachedExtent &extent) final {
+    auto s = extent.get_2q_state();
+    if (extent.is_linked_to_list()) {
+      if (s == extent_2q_state_t::WarmIn) {
+	warm_in.remove(extent);
+      } else {
+	ceph_assert(s == extent_2q_state_t::Hot);
+	hot.remove(extent);
+      }
+    } else {
+      ceph_assert(s == extent_2q_state_t::Fresh);
+    }
+  }
+
+  void move_to_top(
+    CachedExtent &extent,
+    const Transaction::src_t* p_src,
+    extent_len_t load_start) final {
+    auto state = extent.get_2q_state();
+    auto type = extent.get_type();
+    auto loaded_len =
+	extent.is_fully_loaded()
+	? extent.get_length()
+	: extent.get_loaded_length();
+    if (extent.is_linked_to_list()) {
+      if (state == extent_2q_state_t::Hot) {
+	hot.move_to_top(extent, p_src);
+	hit_queue(overall_hits.hot_hits, p_src, type, loaded_len);
+      } else {
+	ceph_assert(state == extent_2q_state_t::WarmIn);
+	hit_queue(overall_hits.warm_in_hits, p_src, type, loaded_len);
+	// warm_in is a FIFO queue, do nothing here
```

Comment:
Let's emphasize in comment here that in the standard 2Q algorithm, the extent won't be considerred hot until it is evicted to the warm out queue.